allennlp/data/dataset_readers/quora_paraphrase.py

```diff
@@ -0,0 +1,82 @@
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+def parse_file_uri(uri: str):
+    match = re.fullmatch(r'\((.*)\)#(.*)', uri)      # pylint: disable=anomalous-backslash-in-string
+    if match and len(match.groups()) == 2:
+        return match.groups()[0], match.groups()[1]
+    else:
+        return uri, None
+
+
+@DatasetReader.register("quora_paraphrase")
+class QuoraParaphraseDatasetReader(DatasetReader):
+    """
+    Reads a file from the Quora Paraphrase dataset. The train/validation/test split of the data
+    comes from the paper `Bilateral Multi-Perspective Matching for Natural Language Sentences
+    <https://arxiv.org/abs/1702.03814>`_ by Zhiguo Wang et al., 2017. Each file of the data
+    is a tsv file without header. The columns are is_duplicate, question1, question2, and id.
+    All questions are pre-tokenized and tokens are space separated. We convert these keys into
+    fields named "label", "premise" and "hypothesis", so that it is compatible to some existing
+    natural language inference algorithms.
+
+    Parameters
+    ----------
+    lazy : ``bool`` (optional, default=False)
+        Passed to ``DatasetReader``.  If this is ``True``, training will start sooner, but will
+        take longer per batch.  This also allows training with datasets that are too large to fit
+        in memory.
+    tokenizer : ``Tokenizer``, optional
+        Tokenizer to use to split the premise and hypothesis into words or other kinds of tokens.
+        Defaults to ``WordTokenizer(JustSpacesWordSplitter())``.
+    token_indexers : ``Dict[str, TokenIndexer]``, optional
+        Indexers used to define input token representations. Defaults to ``{"tokens":
+        SingleIdTokenIndexer()}``.
+    """
+    def __init__(self,
+                 lazy: bool = False,
+                 tokenizer: Tokenizer = None,
+                 token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        super().__init__(lazy)
+        self._tokenizer = tokenizer or WordTokenizer(JustSpacesWordSplitter())
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+
+    @overrides
+    def _read(self, file_path):
+        logger.info("Reading instances from lines in file at: %s", file_path)
+        with open(cached_path(file_path), "r") as data_file:
+            tsv_in = csv.reader(data_file, delimiter='\t')
+            for row in tsv_in:
+                if len(row) == 4:
+                    yield self.text_to_instance(premise=row[1], hypothesis=row[2], label=row[0])
+
+    @overrides
+    def text_to_instance(self,  # type: ignore
+                         premise: str,
+                         hypothesis: str,
+                         label: str = None) -> Instance:
+        # pylint: disable=arguments-differ
+        fields: Dict[str, Field] = {}
+        tokenized_premise = self._tokenizer.tokenize(premise)
```

Comment:
Oh, these are pre-tokenized!  That's an important point, thanks for adding it to the class docstring.  I think that means you should _not_ tokenize the text here.  The reason is because in a demo, you'll want to do tokenization (because someone typing text into a demo won't necessarily tokenize it nicely for you).  So if/when we build a demo for this, we'll want the `Predictor` to have a tokenizer and pass pre-tokenized text into here.  So I'd recommend having `premise: List[str]` and `hypothesis: List[str]`, removing the tokenizer, and just calling `row[1].split()` and `row[2].split()` above.  Does that make sense?

Reply:
Sure it makes sense. But as text_to_instance is an overridden method, will the change of types cause any side effect? How about we do this in the next PR where we add the predictor and demo?

Reply:
Also, the raw Quora data is actually not pre-tokenized. Only this particular train/validate/test split is pre-tokenized. At sometime I will want to use the same split but use non-tokenized version to see if better tokenization could further improve the performance.

Reply:
Oh, ok, that's a good point.  And yeah, it's fine to worry about this in a subsequent PR.  It's just something to be careful about when this actually goes into a demo.

Reply:
Oh, and for the types, we're already ignoring types on these methods, because none of them agree with the super class.  The point of that method existing on the super class is just to tell you that you really should implement a method with this name, though the particulars of how it looks are going to be specific to your data.